### PR TITLE
[ Feat ] Validate whisky.json

### PR DESCRIPTION
**Pull request by gpibarra**
I was modifying the `whisky.json` file and by mistake generated an invalid json file (a trailing comma).
When running `whisky update`, the error was not clear enough, because doing `json_decode()` did not check if the file was well formed and returned null
From there, I decided to also implement the validation of all json properties to avoid future errors.

PS: need to do the build.

### CODE REVIEW

File: app/FileJson.php [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+            } catch (\Exception $e) {
+                $msg = 'Invalid JSON schema';
+                $msg .= ': '.$e->getMessage();
+                $msg .= ' in '.$this->path;
+                throw new \Exception($msg);
+            }
+        }
+
+        return $content;
+    }
```
Comment (faissaloux) [inline app/FileJson.php:50]:
```suggestion
    public function read(bool $validate = true): string|array|null
    {
        if (json_last_error() !== JSON_ERROR_NONE) {
            $msg = 'Invalid JSON';
            if (function_exists('json_last_error_msg')) {
                $msg .= ': '.json_last_error_msg();
                $msg .= ' in '.$this->path;
            }
            throw new \Exception($msg);
        }

        if ($validate) {
            try {
                $contentJson = json_decode(File::get($this->path));
                $this->validate($contentJson);
            } catch (\Exception $e) {
                $msg = 'Invalid JSON schema';
                $msg .= ': '.$e->getMessage();
                $msg .= ' in '.$this->path;
                throw new \Exception($msg);
            }
        }

        return json_decode(File::get($this->path), true);
    }
```

File: app/FileJson.php [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+        // $options->version = 7;
+
+        $schema = Schema::import($this->getSchemaValidation(), $options);
+
+        $schema->in($content);
+    }
+
+    protected function getSchemaValidation(): ?\stdClass
+    {
+        // return json_decode(File::get(Whisky::base_path('resources/schemas/whisky.json')), false);
```
Comment (faissaloux) [inline app/FileJson.php:69]:
```suggestion
        
```

**Review by faissaloux — CHANGES_REQUESTED** [review]

**Review by faissaloux — CHANGES_REQUESTED** [review]

**Comment by ProjektGopher** [comment]
This looks super promising! I'm excited to look through it properly when I'm sitting at a computer after I get my kid down for the night

File: app/FileJson.php (hunk 1 above)
Comment (gpibarra) [reply app/FileJson.php:50]: I first validate the `json_last_error` function but `json_decode` must have been executed before. Then I do the validation with the schema library.
Note that the `read` function returns an "array", so it returns `$content` with `json_decode($string, true)`, while schema validation needs a "stdClass" so `$contentJson` is with `json_decode($string, false)`.

File: app/FileJson.php (hunk 2 above)
Comment (ProjektGopher) [reply app/FileJson.php:67]: I don't mind the idea of extracting this, so I'm going to leave this comment in as a reminder for now.

File: app/Whisky.php [hunk 3]
```diff
@@ -50,9 +49,7 @@ public static function base_path(string $path = ''): string
 
     public static function readConfig(string $key): string|array|null
     {
-        $path = self::cwd('whisky.json');
-
-        $cfg = json_decode(File::get($path), true);
+        $cfg = FileJson::make(self::cwd('whisky.json'))->read();
```
Comment (ProjektGopher) [inline app/Whisky.php:52]: I'd like to refactor how this is called eventually to make it a bit cleaner, but for now it gets us validated so I'm going to just leave it.

**Review by ProjektGopher — APPROVED** [review]

**Pull request merged**